Source/cairo/WaylandCairoSurface.m

```diff
@@ -0,0 +1,204 @@
+/*
+   Copyright (C) 2016 Sergio L. Pascual <slp@sinrega.org>
+*/
+
+#define _GNU_SOURCE
+
+#include "wayland/WaylandServer.h"
+#include "cairo/WaylandCairoSurface.h"
+#include <cairo/cairo.h>
+
+#include <stdlib.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <sys/mman.h>
+
+#define GSWINDEVICE ((struct window *)gsDevice)
+
+/* Linux specific version */
+static int
+os_create_anonymous_file(off_t size)
+{
+    static const char template[] = "/weston-shared-XXXXXX";
+    const char *path;
+    char *name;
+    int fd;
+
+    path = getenv("XDG_RUNTIME_DIR");
+    if (!path) {
+	errno = ENOENT;
+	return -1;
+    }
+
+    name = malloc(strlen(path) + sizeof(template));
+    if (!name)
+	return -1;
+
+    strcpy(name, path);
+    strcat(name, template);
+
+    fd = memfd_create(name, MFD_CLOEXEC);
+
+    free(name);
+
+    if (fd < 0)
+	return -1;
+
+    if (ftruncate(fd, size) != 0) {
+	close(fd);
+	return -1;
+    }
+
+    return fd;
+}
+
+static cairo_surface_t *
+create_shm_buffer(struct window *window)
+{
+    struct wl_shm_pool *pool;
+    cairo_surface_t *surface;
+    int fd, size, stride;
+
+    stride = window->width * 4;
+    size = stride * window->height;
+
+    NSDebugLog(@"WaylandCairoSurface: creating shm buffer of %d bytes", size);
+    fd = os_create_anonymous_file(size);
+    if (fd < 0) {
+	NSLog(@"creating a buffer file for surface failed");
+	return NULL;
+    }
+
+    window->data =
+	mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
+    if (window->data == MAP_FAILED) {
+	NSLog(@"error mapping anonymous file");
+	close(fd);
+	return NULL;
+    }
+
+    pool = wl_shm_create_pool(window->wlconfig->shm, fd, size);
+
+    surface = cairo_image_surface_create_for_data(window->data,
+						  CAIRO_FORMAT_ARGB32,
+						  window->width,
+						  window->height,
+						  stride);
+
+    window->buffer =
+	wl_shm_pool_create_buffer(pool, 0,
+				  window->width, window->height, stride,
+				  WL_SHM_FORMAT_ARGB8888);
+    wl_shm_pool_destroy(pool);
+
+    close(fd);
+
+    return surface;
+}
+
+@implementation WaylandCairoSurface
+
+- (id) initWithDevice: (void*)device
+{
+    struct window *window = (struct window *) device;
+    NSDebugLog(@"WaylandCairoSurface: initWithDevice win=%d", window->window_id);
+
+    gsDevice = device;
+
+    _surface = create_shm_buffer(window);
+    if (_surface == NULL) {
+	NSDebugLog(@"can't create cairo surface");
+	return 0;
+    }
+
+    wl_surface_attach(window->surface, window->buffer, 0, 0);
+    window->wcs = self;
+
+    return self;
+}
+
+- (void) dealloc
+{
+    struct window *window = (struct window*) gsDevice;
+    NSDebugLog(@"WaylandCairoSurface: dealloc win=%d", window->window_id);
```

Comment:
Shouldn't this free the shm buffer?

Reply:
Attempting to find out exactly what needs to be freed is proving to be tricky.

`munmap()`ing the memory seemingly used by the Cairo surface without freeing the Cairo surface seems wrong to me, and calling `cairo_surface_destroy()` here to decrease the refcount and likely destroy the surface seems to be causing crash within `libwayland-client` code.

I'm therefore leaving an extensive comment here and marking this resolved.